soc/intel/intel_adsp/common/debug_window.c

```diff
@@ -0,0 +1,158 @@
+/* Copyright (c) 2025 Intel Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <adsp_memory.h>
+#include <adsp_debug_window.h>
+#include <zephyr/cache.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(debug_window, CONFIG_SOC_LOG_LEVEL);
+
+struct adsp_debug_window {
+	struct adsp_dw_desc descs[ADSP_DW_DESC_COUNT];
+	uint8_t reserved[ADSP_DW_PAGE0_SLOT_OFFSET -
+			 ADSP_DW_DESC_COUNT * sizeof(struct adsp_dw_desc)];
+	uint8_t partial_page0[ADSP_DW_SLOT_SIZE - ADSP_DW_PAGE0_SLOT_OFFSET];
+	uint8_t slots[ADSP_DW_SLOT_COUNT][ADSP_DW_SLOT_SIZE];
+} __packed;
+
+#define WIN2_MBASE		DT_REG_ADDR(DT_PHANDLE(DT_NODELABEL(mem_window2), memory))
+#define WIN2_SLOTS		((HP_SRAM_WIN2_SIZE / ADSP_DW_PAGE_SIZE) - 1)
+#define ADSP_DW_FULL_SLOTS	MIN(WIN2_SLOTS, ADSP_DW_SLOT_COUNT)
+
+#define ADSP_DW ((volatile struct adsp_debug_window *) \
+		  (sys_cache_uncached_ptr_get((__sparse_force void __sparse_cache *) \
+				     (WIN2_MBASE + WIN2_OFFSET))))
+
+static int adsp_dw_find_slot(uint32_t type, bool prefer_partial)
+{
+	int i;
+
+	if (prefer_partial) {
+		if (ADSP_DW->descs[ADSP_DW_SLOT_COUNT].type == type) {
+			return ADSP_DW_SLOT_COUNT;
+		}
+	}
+
+	for (i = 0; i < ADSP_DW_FULL_SLOTS; i++) {
+		if (ADSP_DW->descs[i].type == type) {
+			return i;
+		}
+	}
+
+	if (!prefer_partial) {
+		if (ADSP_DW->descs[ADSP_DW_SLOT_COUNT].type == type) {
+			return ADSP_DW_SLOT_COUNT;
+		}
+	}
+
+	return -ENOENT;
+}
+
+volatile void *adsp_dw_request_slot(struct adsp_dw_desc *dw_desc, size_t *slot_size)
+{
+	int slot_idx;
+
+	if (!dw_desc->type) {
+		return NULL;
+	}
+
+	/* Check if a slot have been allocated for this type */
+	slot_idx = adsp_dw_find_slot(dw_desc->type, !!slot_size);
+	if (slot_idx < 0) {
+		/* Look for an empty slot */
+		slot_idx = adsp_dw_find_slot(ADSP_DW_SLOT_UNUSED, !!slot_size);
+		if (slot_idx < 0) {
+			LOG_INF("No available slot for %#x", dw_desc->type);
+			return NULL;
+		}
+	}
+
+	ADSP_DW->descs[slot_idx].resource_id = dw_desc->resource_id;
```

Comment:
So if adsp_dw_find_slot returns a valid slot_idx, this function will still overtake this slot and overwrite resource_id and vma even if the previous user did not release it? Is this intentional? Because looks like a task intended for adsp_dw_seize_slot.